### PR TITLE
Introducing IConnection interface

### PR DESCRIPTION
**Pull request by moozzyk**

### CODE REVIEW

File: src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs [hunk 1]
```diff
@@ -46,15 +46,20 @@ public class HubConnection
...
+
+        public HubConnection(IConnection connection, IInvocationAdapter adapter, ILoggerFactory loggerFactory)
         {
             // TODO: loggerFactory shouldn't be required
             if (loggerFactory == null)
             {
                 throw new ArgumentNullException(nameof(loggerFactory));
             }
 
+            _connection = connection ?? throw new ArgumentNullException(nameof(connection));
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs]: No throw expressions. I hate these things.

File: test/Microsoft.AspNetCore.Sockets.Client.Tests/HubConnectionTests.cs [hunk 2]
```diff
@@ -167,5 +165,62 @@ public void CannotCreateHubConnectionWithNullUrl()
...
+            var mockConnection = new Mock<IConnection>();
+            mockConnection
+                .Setup(m => m.DisposeAsync())
+                .Callback(() => mockConnection.Raise(c => c.Closed += null, (Exception)null))
+                .Returns(Task.FromResult<object>(null));
+
+            var hubConnection = new HubConnection(mockConnection.Object, Mock.Of<IInvocationAdapter>(), new LoggerFactory());
+
+            await hubConnection.StartAsync(Mock.Of<ITransport>());
+            var invokeTask = hubConnection.Invoke("testMethod", typeof(int));
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.Sockets.Client.Tests/HubConnectionTests.cs]: Is this racey? Invoke might finish before DisposeAsync?

File: src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs (hunk 1 above)
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs]:
```
if(((_connection = connection) == null) == true) {
    string paramName = default(string);
    paramName = nameof(connection);
    throw new ArgumentNullException(paramName);
}
```

File: test/Microsoft.AspNetCore.Sockets.Client.Tests/HubConnectionTests.cs (hunk 2 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.Sockets.Client.Tests/HubConnectionTests.cs]: Good catch. Will fix that.

Comment (moozzyk) [inline test/Microsoft.AspNetCore.Sockets.Client.Tests/HubConnectionTests.cs]: Actually, no. There is no server so no one can complete this invocation/invokeTask.

**Pull request closed without merging**